lib/ecto/adapters/sql.ex

```diff
@@ -650,6 +650,9 @@ defmodule Ecto.Adapters.SQL do
       {:ok, %{rows: nil, num_rows: 1}} ->
         {:ok, []}
 
+      {:ok, %{rows: [], num_rows: 1}} ->
+        {:ok, []}
```

Comment:
If the adapter cannot return any rows for an operation, than `rows` must be `nil`. Can you please check this is the case in TDS? If rows is a lost, then it must always have the same amount of entries as `num_rows`.

Reply:
driver will return empty list if:
- colmetadata is present in tds response
- transaction manager response
- prepare

I think I can change that behaviour.

Reply:
if num_rows == 0 it can also return nil?

Reply:
> if num_rows == 0 it can also return nil?

Yes, but if the query would return results when `num_rows > 0`, then it should return an empty list. Here is a postgrex example:

```elixir
iex(1)> Demo.Repo.query("SELECT * FROM users WHERE id = $1", [0])
[debug] QUERY OK db=1.0ms decode=1.9ms queue=1.2ms idle=9545.1ms
SELECT * FROM users WHERE id = $1 [0]
{:ok,
 %Postgrex.Result{
   columns: ["id", "username", "email", "phone_number", "inserted_at",
    "updated_at"],
   command: :select,
   connection_id: 21954,
   messages: [],
   num_rows: 0,
   rows: []
 }}
```

So the question is: can this query return results? If so, you should return a list, otherwise nil. However, it may be this is not an easy scenario to detect in all databases, so here is the minimum I would guarantee:

  * if num_rows == 0, return an empty list
  * if num_rows > 0 and there are no rows,  return nil

Reply:
ok, I will remove this change from sql.ex and try to change tds. I think I can cover the minimum here

Reply:
removed